k2/addons/k2-upgrade/upgrade/upgrader.py

```diff
@@ -0,0 +1,60 @@
+import logging
+from concurrent.futures import TimeoutError
+
+from zaf.extensions.extension import get_logger_name
+
+from upgrade import PERFORM_UPGRADE, UPGRADE_DEFAULT_TIMEOUT
+
+logger = logging.getLogger(get_logger_name('k2', 'upgrade'))
+
+
+class NoUpgradeTypes(Exception):
+    pass
+
+
+class MultipleUpgradeTypes(Exception):
+    pass
+
+
+class UpgradeFailed(Exception):
+    pass
+
+
+class Upgrader(object):
+
+    def __init__(self, messagebus, entity, receiver_endpoint_id):
+        self.messagebus = messagebus
+        self.entity = entity
+        self.receiver_endpoint_id = receiver_endpoint_id
+
+    def _singular_future_or_raise(self, futures):
+        if len(futures) == 0:
+            msg = 'Upgrade failed due to no upgrade types available for sut {sut}'.format(
+                sut=self.entity)
+            logger.error(msg)
+            raise NoUpgradeTypes(msg)
+        elif len(futures) > 1:
+            msg = 'Upgrade failed due to multiple upgrade types registered for sut {sut}'.format(
+                sut=self.entity)
+            logger.error(msg)
+            raise MultipleUpgradeTypes(msg)
+
+    def start_upgrade(self, software):
+        logger.info('Upgrading sut {sut}'.format(sut=self.entity))
+        futures = self.messagebus.send_request(
+            PERFORM_UPGRADE,
+            receiver_endpoint_id=self.receiver_endpoint_id,
+            entity=self.entity,
+            data=software)
```

Comment:
I would have expected that the upgrade type was sent here as well. Otherwise the upgrade-implementation must extract it from the config with the assumption that it is on the entity - which might change in the future.

Reply:
I was thinking to have separate upgrade implementations being activated based on the config. Similar how there's only 1 power switch implementation allowed at any one time (for each suts id). I don't expect to have a "1 upgrade addon to catch them all", if that's what you were aiming at. Besides, people might want to have their own custom upgrade addons. I'm currently working on an addon that supports https://mender.io/
Or did you mean that all upgrade addons to be activated always, and only the addons that support the upgrade type will actually do anything with the message?

Reply:
I think I start to understand what you are going for but it is still a bit cloudy.
I think I would have to see what an implementation and example config would look like to fully understand the intent.
If you could write a dummy upgrader as part of system test with a "dummy" upgrade-method to illustrate how the configured upgrade-method, active plugins and configured entities would tie into the whole mechanism would surely help me understand.

Reply:
I'll see what I can do :wink: I'm currently focusing on designing a multiple serial port component, but once I get that working, I can put some more effort into this pull request. For now, I'll leave it as a draft